File: spring_tofu_app/src/main/java/com/trianaSalesianos/tofuApp/model/Recipe.java
package com.trianaSalesianos.tofuApp.model;

import lombok.*;
import org.hibernate.annotations.GenericGenerator;
import org.springframework.data.annotation.CreatedDate;

import javax.persistence.*;
import java.io.Serializable;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
@Entity
@NoArgsConstructor @AllArgsConstructor
@Getter @Setter
@Builder
public class Recipe implements Serializable {
    @Id
    @GeneratedValue(generator = "UUID")
    @GenericGenerator(
            name = "UUID",
            strategy = "org.hibernate.id.UUIDGenerator",
            parameters = {
                    @org.hibernate.annotations.Parameter(
                            name = "uuid_gen_strategy_class",
                            value = "org.hibernate.id.uuid.CustomVersionOneStrategy"
                    )
            }
    )
    private UUID id;
    private String name;
    private String description;

    @Builder.Default
    private Integer prepTime = 0;   // in minutes
    private String steps;
    @Builder.Default
    private String img = "https://ichef.bbci.co.uk/food/ic/food_16x9_832/recipes/british_shakshuka_26737_16x9.jpg";
    private String category;  //Vegetariano, vegano, hiper-proteico, hiper-calorico, hipo-calorico...
    @Builder.Default
    private LocalDateTime createdAt = LocalDateTime.now();

    //TODO Gestion de fetch type lazy con subgraphos y seteo por defecto de las listas en vacia
    //TODO Gestion del borrado de los ingredientes y usuario
    // (si se borra receta, no se borra usuario, pero si se borra usuario se borra receta)
    @ToString.Exclude
    @ManyToOne()
    @JoinColumn(name = "user_id", foreignKey = @ForeignKey(name = "FK_RECIPE_USER"))
    private User author;
    @ToString.Exclude
    @Builder.Default
    @ManyToMany(mappedBy = "favorites", fetch = FetchType.EAGER)
    private List<User> favoritedBy = new ArrayList<>();

    @Builder.Default
    @OneToMany(mappedBy = "recipe")
    private List<RecipeIngredient> recipeIngredients = new ArrayList<>();
}

File: flutter_tofu_app/lib/models/recipe.dart
class RecipeResponse {
  final String? id;
  final String? name;
  final String? description;
  final String? category;
  final String? img;
  final String? author;
  final int? prepTime;
  final String? createdAt;
  const RecipeResponse(
      {this.id,
      this.name,
      this.description,
      this.category,
      this.img,
      this.author,
      this.prepTime,
      this.createdAt});
  RecipeResponse copyWith(
      {String? id,
      String? name,
      String? description,
      String? category,
      String? img,
      String? author,
      int? prepTime,
      String? createdAt}) {
    return RecipeResponse(
        id: id ?? this.id,
        name: name ?? this.name,
        description: description ?? this.description,
        category: category ?? this.category,
        img: img ?? this.img,
        author: author ?? this.author,
        prepTime: prepTime ?? this.prepTime,
        createdAt: createdAt ?? this.createdAt);
  }

  Map<String, Object?> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'category': category,
      'img': img,
      'author': author,
      'prepTime': prepTime,
      'createdAt': createdAt
    };
  }

  static RecipeResponse fromJson(Map<String, Object?> json) {
    return RecipeResponse(
        id: json['id'] == null ? null : json['id'] as String,
        name: json['name'] == null ? null : json['name'] as String,
        description:
            json['description'] == null ? null : json['description'] as String,
        category: json['category'] == null ? null : json['category'] as String,
        img: json['img'] == null ? null : json['img'] as String,
        author: json['author'] == null ? null : json['author'] as String,
        prepTime: json['prepTime'] == null ? null : json['prepTime'] as int,
        createdAt:
            json['createdAt'] == null ? null : json['createdAt'] as String);
  }

  @override
  String toString() {
    return '''RecipeResponse(
                id:$id,
                name:$name,
                description:$description,
                category:$category,
                img:$img,
                author:$author,
                prepTime:$prepTime,
                createdAt:$createdAt
    ) ''';
  }

  @override
  bool operator ==(Object other) {
    return other is RecipeResponse &&
        other.runtimeType == runtimeType &&
        other.id == id &&
        other.name == name &&
        other.description == description &&
        other.category == category &&
        other.img == img &&
        other.author == author &&
        other.prepTime == prepTime &&
        other.createdAt == createdAt;
  }

  @override
  int get hashCode {
    return Object.hash(runtimeType, id, name, description, category, img,
        author, prepTime, createdAt);
  }
}
